src/collector/rrdp.rs

```diff
@@ -1091,6 +1104,98 @@ impl HttpClient {
 }
 
 
+//------------ HttpResponse --------------------------------------------------
+
+/// Wraps a reqwest response for added features.
+struct HttpResponse {
+    /// The wrapped reqwest response.
+    response: Response,
+
+    /// A file to also store read data into.
+    file: Option<fs::File>,
+}
+
+impl HttpResponse {
+    pub fn create(
+        response: Response,
+        uri: &uri::Https,
+        response_dir: &Option<PathBuf>,
+        multi: bool
+    ) -> Self {
+        HttpResponse {
+            response,
+            file: response_dir.as_ref().and_then(|base| {
+                Self::open_file(base, uri, multi).ok()
+            })
+        }
+    }
+
+    fn open_file(
+        base: &Path, uri: &uri::Https, multi: bool
+    ) -> Result<fs::File, Failed> {
+        let path = base.join(&uri.as_str()[8..]);
+        let path = if multi {
+            path.join(Utc::now().to_rfc3339())
+        }
+        else {
+            path
+        };
+
+        let parent = match path.parent() {
+            Some(parent) => parent,
+            None => {
+                warn!(
+                    "Cannot keep HTTP response; \
+                    URI translated into a bad path '{}'",
+                    path.display()
+                );
+                return Err(Failed)
+            }
+        };
+        if let Err(err) = fs::create_dir_all(&parent) {
+            warn!(
+                "Cannot keep HTTP response; \
+                creating director {} failed: {}",
+                parent.display(), err
+            );
+            return Err(Failed)
+        }
+        fs::File::create(&path).map_err(|err| {
+            warn!(
+                "Cannot keep HTTP response; \
+                creating file {} failed: {}",
+                path.display(), err
+            );
+            Failed
+        })
+    }
+
+    pub fn content_length(&self) -> Option<u64> {
+        self.response.content_length()
+    }
+
+    pub fn copy_to<W: io::Write + ?Sized>(
+        &mut self, w: &mut W
+    ) -> Result<u64, io::Error> {
+        io::copy(self, w)
```

Comment:
Consider adding a comment explaining that the `fn read` in `impl io::Read for HttpResponse` is what actually decides what is being written here because of the Reader/Writer interplay in `io::copy` (i.e. only the `.response` from the HttpResponse struct, not the entire struct including the optional `.file`).